include/deal.II/base/aligned_vector.h

```diff
@@ -1036,6 +1124,418 @@ AlignedVector<T>::fill(const T &value)
 
 
 
+template <class T>
+inline void
+AlignedVector<T>::replicate_across_communicator(const MPI_Comm &   communicator,
+                                                const unsigned int root_process)
+{
+#  ifdef DEAL_II_WITH_MPI
+#    if DEAL_II_MPI_VERSION_GTE(3, 0)
+
+  // **** Step 0 ****
+  // All but the root process no longer need their data, so release the memory
+  // used to store the previous elements.
+  if (Utilities::MPI::this_mpi_process(communicator) != root_process)
+    {
+      elements.reset();
+      used_elements_end      = nullptr;
+      allocated_elements_end = nullptr;
+    }
+
+  // **** Step 1 ****
+  // Create communicators for each group of processes that can use
+  // shared memory areas. Within each of these groups, we don't care about
+  // which rank each of the old processes gets except that we would like to
+  // make sure that the (global) root process will have rank=0 within
+  // its own sub-communicator. We can do that through the third argument of
+  // MPI_Comm_split_type (the "key") which is an integer meant to indicate the
+  // order of processes within the split communicators, and we will set it to
+  // zero for the root processes and one for all others -- which means that
+  // for all of these other processes, MPI can choose whatever order it
+  // wants because they have the same key (MPI then documents that these ties
+  // will be broken according to these processes' rank in the old group).
+  //
+  // At least that's the theory. In practice, the MPI implementation where
+  // this function was developed on does not seem to do that. But it
+  // is willing to put the root process onto the *last* rank of the
+  // communicator.
```

Comment:
The function `MPI_Comm_split_type` seems to ignore for some reason the key. What one could do is to call `MPI_Comm_split_type` without any keys and resort the ranks - if needed - afterwards.

```diff
diff --git a/include/deal.II/base/aligned_vector.h b/include/deal.II/base/aligned_vector.h
index 9aab94e..b16be26 100644
--- a/include/deal.II/base/aligned_vector.h
+++ b/include/deal.II/base/aligned_vector.h
@@ -1161,26 +1161,32 @@ AlignedVector<T>::replicate_across_communicator(const MPI_Comm &   communicator,
   // communicator.
   MPI_Comm shmem_group_communicator;
   {
+    MPI_Comm shmem_group_communicator_temp;
+    int      ierr =
+      MPI_Comm_split_type(communicator,
+                          MPI_COMM_TYPE_SHARED,
+                          Utilities::MPI::this_mpi_process(communicator),
+                          MPI_INFO_NULL,
+                          &shmem_group_communicator_temp);
+
+    AssertThrowMPI(ierr);
     const int key =
       (Utilities::MPI::this_mpi_process(communicator) == root_process ?
-         Utilities::MPI::n_mpi_processes(communicator) :
-         0);
-    const int ierr = MPI_Comm_split_type(communicator,
-                                         MPI_COMM_TYPE_SHARED,
-                                         key,
-                                         MPI_INFO_NULL,
-                                         &shmem_group_communicator);
+         0 :
+         Utilities::MPI::this_mpi_process(communicator) + 1);
+    ierr = MPI_Comm_split(shmem_group_communicator_temp,
+                          1,
+                          key,
+                          &shmem_group_communicator);
     AssertThrowMPI(ierr);
 
     // Verify the explanation from above
     if (Utilities::MPI::this_mpi_process(communicator) == root_process)
-      Assert(Utilities::MPI::this_mpi_process(shmem_group_communicator) ==
-               Utilities::MPI::n_mpi_processes(shmem_group_communicator) - 1,
+      Assert(Utilities::MPI::this_mpi_process(shmem_group_communicator) == 0,
              ExcInternalError());
   }
   const bool is_shmem_root =
-    Utilities::MPI::this_mpi_process(shmem_group_communicator) ==
-    Utilities::MPI::n_mpi_processes(shmem_group_communicator) - 1;
+    Utilities::MPI::this_mpi_process(shmem_group_communicator) == 0;
 
   // **** Step 2 ****
   // We then have to send the state of the current object from the
@@ -1216,8 +1222,8 @@ AlignedVector<T>::replicate_across_communicator(const MPI_Comm &   communicator,
   {
     const int key =
       (Utilities::MPI::this_mpi_process(communicator) == root_process ?
-         Utilities::MPI::n_mpi_processes(communicator) :
-         0);
+         0 :
+         Utilities::MPI::this_mpi_process(communicator) + 1);
 
     const int ierr = MPI_Comm_split(communicator,
                                     /*color=*/
@@ -1228,13 +1234,11 @@ AlignedVector<T>::replicate_across_communicator(const MPI_Comm &   communicator,
 
     // Again verify the explanation from above
     if (Utilities::MPI::this_mpi_process(communicator) == root_process)
-      Assert(Utilities::MPI::this_mpi_process(shmem_roots_communicator) ==
-               Utilities::MPI::n_mpi_processes(shmem_roots_communicator) - 1,
+      Assert(Utilities::MPI::this_mpi_process(shmem_roots_communicator) == 0,
              ExcInternalError());
   }
-  const unsigned int shmem_roots_root_rank =
-    Utilities::MPI::n_mpi_processes(shmem_roots_communicator) - 1;
-  const bool is_shmem_roots_root =
+  const unsigned int shmem_roots_root_rank = 0;
+  const bool         is_shmem_roots_root =
     (Utilities::MPI::this_mpi_process(shmem_roots_communicator) ==
      shmem_roots_root_rank);
 
@@ -1260,12 +1264,11 @@ AlignedVector<T>::replicate_across_communicator(const MPI_Comm &   communicator,
             resize(new_size);
 
           // Then directly copy from the root process into these buffers
-          int ierr = MPI_Bcast(
-            elements.get(),
-            sizeof(T) * new_size,
-            MPI_CHAR,
-            Utilities::MPI::n_mpi_processes(shmem_roots_communicator) - 1,
-            shmem_roots_communicator);
+          int ierr = MPI_Bcast(elements.get(),
+                               sizeof(T) * new_size,
+                               MPI_CHAR,
+                               0,
+                               shmem_roots_communicator);
           AssertThrowMPI(ierr);
         }
       else
@@ -1279,8 +1282,7 @@ AlignedVector<T>::replicate_across_communicator(const MPI_Comm &   communicator,
           // pointer, which would trigger the deleter which would lead to a
           // deadlock. So we just send the result of the broadcast() call to
           // nirvana on the root process and keep our current state.
-          if (Utilities::MPI::this_mpi_process(shmem_roots_communicator) ==
-              Utilities::MPI::n_mpi_processes(shmem_roots_communicator) - 1)
+          if (Utilities::MPI::this_mpi_process(shmem_roots_communicator) == 0)
             Utilities::MPI::broadcast(shmem_roots_communicator,
                                       *this,
                                       shmem_roots_root_rank);
@@ -1340,11 +1342,11 @@ AlignedVector<T>::replicate_across_communicator(const MPI_Comm &   communicator,
   // able to learn about the array size that way.
   MPI_Win        shmem_window;
   void *         base_ptr;
-  const MPI_Aint align_by   = 64;
-  const MPI_Aint alloc_size = Utilities::MPI::broadcast(
-    shmem_group_communicator,
-    (size() * sizeof(T) + align_by - 1),
-    Utilities::MPI::n_mpi_processes(shmem_group_communicator) - 1);
+  const MPI_Aint align_by = 64;
+  const MPI_Aint alloc_size =
+    Utilities::MPI::broadcast(shmem_group_communicator,
+                              (size() * sizeof(T) + align_by - 1),
+                              0);
 
   {
     const int disp_unit = (is_shmem_root ? size() : 1);

````
This seems to work.

Reply:
Right, but does this actually improve things? I would (i) still have to explain in the comments why I'm not doing the renumbering in `MPI_Comm_split_type` lest someone come along and make that change, (ii) we're adding another collective communication step.

I don't disagree that one *could* do it, I just don't quite see *why*. But I'm also not opposed if you're adamant about it.

Reply:
Not making rank 0 to the root look a bit odd. That the `keys` do not work properly is a bit worrying. I am fine to leave it as it is but have you tried to run the tests also on multiple compute nodes? Is the behavior the same there? The tests in the test suite can only test the case that all processes are on the same shmem domain. This will probably not change as long as not a user-provided communicator can passed the function.